File: packages/types/src/material.ts
/**
 * 物料相关类型定义
 */

import type { BaseEntity, StatusEntity, TreeNode } from './base';

/** 库存状态 */
export type StockStatus = 'in_stock' | 'out_of_stock';

/** 审核状态 */
export type AuditStatus = 'pending' | 'approved' | 'rejected';

/** 物料分类 */
export interface Category extends StatusEntity {
  /** 分类名称 */
  name: string;
  /** 分类图标URL */
  icon?: string;
  /** 排序值 */
  sortOrder: number;
  /** 父分类ID */
  parentId?: number;
  /** 层级(1/2/3) */
  level: number;
  /** 路径(如1/2/3) */
  path: string;
  /** 加价开关 */
  markupEnabled: boolean;
  /** 子分类列表 */
  children?: Category[];
}

/** 分类树节点 */
export type CategoryTreeNode = TreeNode<Category>;

/** 物料基础信息 */
export interface Material extends StatusEntity {
  /** 物料编号 */
  materialNo: string;
  /** 分类ID */
  categoryId: number;
  /** 物料通用名称 */
  name: string;
  /** 物料别名 */
  alias?: string;
  /** 物料描述 */
  description?: string;
  /** 默认图片 */
  imageUrl?: string;
  /** 搜索关键词 */
  keywords?: string;
  /** 排序值 */
  sortOrder: number;
  /** 关联分类 */
  category?: Category;
  /** SKU列表 */
  skus?: MaterialSku[];
}

/** 物料SKU */
export interface MaterialSku extends StatusEntity {
  /** SKU编号 */
  skuNo: string;
  /** 物料ID */
  materialId: number;
  /** 品牌 */
  brand: string;
  /** 规格 */
  spec: string;
  /** 销售单位 */
  unit: string;
  /** 重量(kg) */
  weight?: number;
  /** 条形码 */
  barcode?: string;
  /** SKU专属图片 */
  imageUrl?: string;
  /** 关联物料 */
  material?: Material;
  /** 供应商报价列表 */
  supplierMaterials?: SupplierMaterial[];
}

/** 供应商物料报价 */
export interface SupplierMaterial extends StatusEntity {
  /** 供应商ID */
  supplierId: number;
  /** 关联SKU ID */
  materialSkuId: number;
  /** 供应商报价 */
  price: number;
  /** 原价（用于显示划线价） */
  originalPrice?: number;
  /** 最小起订量 */
  minQuantity: number;
  /** 步进数量 */
  stepQuantity: number;
  /** 库存状态 */
  stockStatus: StockStatus;
  /** 审核状态 */
  auditStatus: AuditStatus;
  /** 审核驳回原因 */
  rejectReason?: string;
  /** 是否推荐 */
  isRecommended: boolean;
  /** 销量统计 */
  salesCount: number;
  /** 关联SKU */
  materialSku?: MaterialSku;
  /** 供应商信息 */
  supplier?: {
    id: number;
    name: string;
    displayName?: string;
    minOrderAmount: number;
    deliveryDays: number[];
  };
}

/** 创建分类请求 */
export interface CreateCategoryRequest {
  /** 分类名称 */
  name: string;
  /** 父分类ID */
  parentId?: number;
  /** 分类图标 */
  icon?: string;
  /** 排序值 */
  sortOrder?: number;
  /** 加价开关 */
  markupEnabled?: boolean;
}

/** 更新分类请求 */
export interface UpdateCategoryRequest {
  /** 分类名称 */
  name?: string;
  /** 分类图标 */
  icon?: string;
  /** 排序值 */
  sortOrder?: number;
  /** 加价开关 */
  markupEnabled?: boolean;
  /** 状态 */
  status?: 0 | 1;
}

/** 创建物料请求 */
export interface CreateMaterialRequest {
  /** 分类ID */
  categoryId: number;
  /** 物料名称 */
  name: string;
  /** 物料别名 */
  alias?: string;
  /** 物料描述 */
  description?: string;
  /** 默认图片 */
  imageUrl?: string;
  /** 搜索关键词 */
  keywords?: string;
  /** 排序值 */
  sortOrder?: number;
}

/** 创建SKU请求 */
export interface CreateMaterialSkuRequest {
  /** 物料ID */
  materialId: number;
  /** 品牌 */
  brand: string;
  /** 规格 */
  spec: string;
  /** 销售单位 */
  unit: string;
  /** 重量 */
  weight?: number;
  /** 条形码 */
  barcode?: string;
  /** SKU图片 */
  imageUrl?: string;
}

/** 创建供应商物料报价请求 */
export interface CreateSupplierMaterialRequest {
  /** SKU ID */
  materialSkuId: number;
  /** 报价 */
  price: number;
  /** 原价 */
  originalPrice?: number;
  /** 最小起订量 */
  minQuantity?: number;
  /** 步进数量 */
  stepQuantity?: number;
  /** 库存状态 */
  stockStatus?: StockStatus;
}

/** 更新供应商物料报价请求 */
export interface UpdateSupplierMaterialRequest {
  /** 报价 */
  price?: number;
  /** 原价 */
  originalPrice?: number;
  /** 最小起订量 */
  minQuantity?: number;
  /** 步进数量 */
  stepQuantity?: number;
  /** 库存状态 */
  stockStatus?: StockStatus;
}

/** 物料查询参数 */
export interface MaterialQueryParams {
  /** 分类ID */
  categoryId?: number;
  /** 供应商ID */
  supplierId?: number;
  /** 搜索关键词 */
  keyword?: string;
  /** 品牌 */
  brand?: string;
  /** 库存状态 */
  stockStatus?: StockStatus;
  /** 是否只显示推荐 */
  isRecommended?: boolean;
}

/** 供应商物料审核请求 */
export interface AuditSupplierMaterialRequest {
  /** 审核状态 */
  auditStatus: 'approved' | 'rejected';
  /** 驳回原因 */
  rejectReason?: string;
}

/** 门店端物料展示（带报价对比） */
export interface MaterialWithPricing extends Material {
  /** 供应商报价列表 */
  prices: SupplierPricing[];
  /** 最低价 */
  minPrice?: number;
  /** 最高价 */
  maxPrice?: number;
}

/** 供应商报价信息 */
export interface SupplierPricing {
  /** 供应商ID */
  supplierId: number;
  /** 供应商名称 */
  supplierName: string;
  /** SKU ID */
  skuId: number;
  /** 品牌 */
  brand: string;
  /** 规格 */
  spec: string;
  /** 单位 */
  unit: string;
  /** 报价 */
  price: number;
  /** 加价后价格 */
  finalPrice: number;
  /** 加价金额 */
  markupAmount: number;
  /** 原价 */
  originalPrice?: number;
  /** 最小起订量 */
  minQuantity: number;
  /** 步进数量 */
  stepQuantity: number;
  /** 库存状态 */
  stockStatus: StockStatus;
  /** 起送价 */
  minOrderAmount: number;
  /** 配送日 */
  deliveryDays: number[];
  /** 图片 */
  imageUrl?: string;
}

File: apps/backend/types/material.go
package types

import "time"

// StockStatus 库存状态枚举
type StockStatus string

const (
	StockInStock    StockStatus = "in_stock"
	StockOutOfStock StockStatus = "out_of_stock"
)

// AuditStatus 审核状态枚举
type AuditStatus string

const (
	AuditPending  AuditStatus = "pending"
	AuditApproved AuditStatus = "approved"
	AuditRejected AuditStatus = "rejected"
)

// CategoryInfo 分类信息
type CategoryInfo struct {
	ID            uint64          `json:"id"`
	Name          string          `json:"name"`
	Icon          string          `json:"icon,omitempty"`
	SortOrder     int             `json:"sortOrder"`
	ParentID      *uint64         `json:"parentId,omitempty"`
	Level         int8            `json:"level"`
	Path          string          `json:"path"`
	MarkupEnabled bool            `json:"markupEnabled"`
	Status        int8            `json:"status"`
	Children      []*CategoryInfo `json:"children,omitempty"`
}

// MaterialInfo 物料基础信息
type MaterialInfo struct {
	ID          uint64        `json:"id"`
	MaterialNo  string        `json:"materialNo"`
	CategoryID  uint64        `json:"categoryId"`
	Name        string        `json:"name"`
	Alias       string        `json:"alias,omitempty"`
	Description string        `json:"description,omitempty"`
	ImageURL    string        `json:"imageUrl,omitempty"`
	Keywords    string        `json:"keywords,omitempty"`
	SortOrder   int           `json:"sortOrder"`
	Status      int8          `json:"status"`
	Category    *CategoryInfo `json:"category,omitempty"`
	Skus        []MaterialSku `json:"skus,omitempty"`
}

// MaterialSku 物料SKU
type MaterialSku struct {
	ID         uint64  `json:"id"`
	SkuNo      string  `json:"skuNo"`
	MaterialID uint64  `json:"materialId"`
	Brand      string  `json:"brand"`
	Spec       string  `json:"spec"`
	Unit       string  `json:"unit"`
	Weight     float64 `json:"weight,omitempty"`
	Barcode    string  `json:"barcode,omitempty"`
	ImageURL   string  `json:"imageUrl,omitempty"`
	Status     int8    `json:"status"`
}

// SupplierMaterialInfo 供应商物料报价信息
type SupplierMaterialInfo struct {
	ID            uint64       `json:"id"`
	SupplierID    uint64       `json:"supplierId"`
	MaterialSkuID uint64       `json:"materialSkuId"`
	Price         float64      `json:"price"`
	OriginalPrice float64      `json:"originalPrice,omitempty"`
	MinQuantity   int          `json:"minQuantity"`
	StepQuantity  int          `json:"stepQuantity"`
	StockStatus   StockStatus  `json:"stockStatus"`
	AuditStatus   AuditStatus  `json:"auditStatus"`
	RejectReason  string       `json:"rejectReason,omitempty"`
	IsRecommended bool         `json:"isRecommended"`
	SalesCount    int          `json:"salesCount"`
	Status        int8         `json:"status"`
	CreatedAt     time.Time    `json:"createdAt"`
	UpdatedAt     time.Time    `json:"updatedAt"`
	Sku           *MaterialSku `json:"sku,omitempty"`
}

// CreateCategoryRequest 创建分类请求
type CreateCategoryRequest struct {
	Name          string  `json:"name" validate:"required,max=50"`
	Icon          string  `json:"icon"`
	SortOrder     int     `json:"sortOrder"`
	ParentID      *uint64 `json:"parentId"`
	MarkupEnabled bool    `json:"markupEnabled"`
}

// UpdateCategoryRequest 更新分类请求
type UpdateCategoryRequest struct {
	Name          string `json:"name" validate:"max=50"`
	Icon          string `json:"icon"`
	SortOrder     int    `json:"sortOrder"`
	MarkupEnabled *bool  `json:"markupEnabled"`
}

// CreateMaterialRequest 创建物料请求
type CreateMaterialRequest struct {
	CategoryID  uint64 `json:"categoryId" validate:"required"`
	Name        string `json:"name" validate:"required,max=100"`
	Alias       string `json:"alias" validate:"max=100"`
	Description string `json:"description"`
	ImageURL    string `json:"imageUrl"`
	Keywords    string `json:"keywords" validate:"max=200"`
	SortOrder   int    `json:"sortOrder"`
}

// CreateMaterialSkuRequest 创建物料SKU请求
type CreateMaterialSkuRequest struct {
	MaterialID uint64  `json:"materialId" validate:"required"`
	Brand      string  `json:"brand" validate:"required,max=50"`
	Spec       string  `json:"spec" validate:"required,max=100"`
	Unit       string  `json:"unit" validate:"required,max=20"`
	Weight     float64 `json:"weight"`
	Barcode    string  `json:"barcode" validate:"max=50"`
	ImageURL   string  `json:"imageUrl"`
}

// CreateSupplierMaterialRequest 创建供应商物料报价请求
type CreateSupplierMaterialRequest struct {
	MaterialSkuID uint64  `json:"materialSkuId" validate:"required"`
	Price         float64 `json:"price" validate:"required,gt=0"`
	OriginalPrice float64 `json:"originalPrice"`
	MinQuantity   int     `json:"minQuantity" validate:"min=1"`
	StepQuantity  int     `json:"stepQuantity" validate:"min=1"`
}

// UpdateSupplierMaterialRequest 更新供应商物料报价请求
type UpdateSupplierMaterialRequest struct {
	Price         float64      `json:"price" validate:"gt=0"`
	OriginalPrice float64      `json:"originalPrice"`
	MinQuantity   int          `json:"minQuantity" validate:"min=1"`
	StepQuantity  int          `json:"stepQuantity" validate:"min=1"`
	StockStatus   *StockStatus `json:"stockStatus"`
}

// MaterialQueryParams 物料查询参数
type MaterialQueryParams struct {
	PaginationQuery
	CategoryID uint64 `json:"categoryId" query:"categoryId"`
	Keyword    string `json:"keyword" query:"keyword"`
	Brand      string `json:"brand" query:"brand"`
}

// SupplierMaterialQueryParams 供应商物料查询参数
type SupplierMaterialQueryParams struct {
	PaginationQuery
	SupplierID  uint64 `json:"supplierId" query:"supplierId"`
	CategoryID  uint64 `json:"categoryId" query:"categoryId"`
	Keyword     string `json:"keyword" query:"keyword"`
	StockStatus string `json:"stockStatus" query:"stockStatus"`
	AuditStatus string `json:"auditStatus" query:"auditStatus"`
}
